qiskit/quantum_info/states/cython/exp_value.pyx

```diff
@@ -0,0 +1,55 @@
+#!python
+#cython: language_level = 3
+#distutils: language = c++
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+cimport cython
+import numpy as np
+
+cpdef unsigned long long popcount(unsigned long long count):
+  count = (count & 0x5555555555555555) + ((count >> 1) & 0x5555555555555555);
+  count = (count & 0x3333333333333333) + ((count >> 2) & 0x3333333333333333);
+  count = (count & 0x0f0f0f0f0f0f0f0f) + ((count >> 4) & 0x0f0f0f0f0f0f0f0f);
+  count = (count & 0x00ff00ff00ff00ff) + ((count >> 8) & 0x00ff00ff00ff00ff);
+  count = (count & 0x0000ffff0000ffff) + ((count >> 16) & 0x0000ffff0000ffff);
+  count = (count & 0x00000000ffffffff) + ((count >> 32) & 0x00000000ffffffff);
+  return count
+
+cpdef double expval_pauli_no_x(complex[:] data, unsigned long long z_mask, complex phase):
+    cpdef double val = 0
+    for i in range(len(data)):
+        current_val = (phase * data[i] * np.conj(data[i])).real
```

Comment:
- `current_val` has no `cdef`.
-  `data[i]*np.conj(data[i])` is just `data.real^2+data.imag^2` which can be simplifed to `data[i].real*data[i].real+data[i].imag*data[i].imag`.  This also avoids a call back into Python